Improve Info HUD with entity information

XMinty77 opened the pull request:
This PR attempts to improve the Info HUD by displaying additional entity information
- Display `TimeActive` & active group for hazard cycles
- Add a way to display the number of frames until the next check for a hazard in Custom Info
- Add `Automatic` mode to Watch Info, which automatically chooses the information to display based on the watched entity type

XMinty77 committed:
feat: add "Mouse" command

XMinty77 committed:
fix: wrong y range check

XMinty77 committed:
fix: wrong range check for mouse position input

XMinty77 committed:
feat: add mouse clicks

added support for mouse clicks and moved mouse position update

DemoJameson committed:
feat: mouse command bug fixes and support x1 x2 and scroll wheel

XMinty77 committed:
feat: draw better prologue end cutscene trigger

draw the trigger area for the bird dash tutorial cutscene at the end of prologue such that it triggers the frame after the bottom right of the hitbox of the player collides with the trigger

XMinty77 committed:
Merge branch 'master' of https://github.com/XMinty7/CelesteTAS

XMinty77 committed:
fix: double render & missing bottom draw

added the missing bottom render for the invincibility trigger box
properly removed the vanilla rendering (double render)

XMinty77 committed:
fix: mishaps in trigger rendering routine

changed the position calculation to use correct values
changed cutscene trigger values for positive trigger height

XMinty77 committed:
refactor: console trigger box renders like vanilla

display improved trigger box when CelesteTAS hitboxes are on, and vanilla trigger box when the debug console is open

DemoJameson committed:
feat: improve the birdnpc hitbox

XMinty77 committed:
Merge branch 'EverestAPI:master' into master

XMinty77 committed:
refactor: improve hazard cycles code

- use the vanilla method Scene.OnInterval to avoid any potential float float behavior inconsistencies from different build modes of the CelesteTAS assembly
- simplify the check for if timeactive ticked
- move the offset finding code and the check distance code into their own public methods for other code to use

XMinty77 committed:
feat: display QoL information for hazard cycles

- display timeactive and the active group
- add a helper method for custom information that allows for displaying number of frames until next check

XMinty77 committed:
feat: auto watch info

- add basic automatic mode for watch info setting, which displays entity-specific information

XMinty77 committed:
refactor: revamp the watch entity system

implements a new system to track watched entities

XMinty77 committed:
refactor: clean up unused code

clean up leftover code from old watch entity tracking system

XMinty77 committed:
fix: mishap in checking watched entities

hitboxes referenced the wrong field WatchingEntities instead of WatchingList to check if an entity is tracked as watched

XMinty77 committed:
fix. watch entity bugs (savestates, no id entities)

fix a mishap where making a savestate, adding a watch entity then loading the savestate removes the watch entity and leaks memory in the checklist
fix lack of support for no id entities in the checklist to be tracked across rooms

XMinty77 committed:
feat: display seeker state & information

XMinty77 committed:
refactor: clean up code

remove pesky empty lines D:<
remove unused methods
add TODO note

XMinty7 commented:
The PR is not finished yet but I removed draft mode so it can be merged to main if needed.

XMinty77 committed:
feat: improve displayed values for seekers

- proper formatting for all values
- fix player angle mishap
- display last player distance and current player distance

XMinty77 committed:
refactor: resolve conflict with upstream

XMinty77 committed:
Merge branch 'master' into master

XMinty77 committed:
Merge branch 'EverestAPI:master' into master

XMinty77 committed:
feat: suppress pause alpha if journal is held

render the Info HUD at full alpha (not transparent) if journal is held while the pause menu is up

XMinty77 committed:
refactor: use player get extension

change Tracker.GetEntity<Player> to the extension method GetPlayer for consistency

XMinty77 committed:
feat: improve watched seeker info

- add draw extensions for crosshair and circle arc
- separated watched entity info code into several files for management, tracking, rendering and auto displaying
- added more and improved current watched seeker information
- added watched seeker rendering
- refactored using directives

XMinty77 committed:
Merge branch 'EverestAPI:master' into master

XMinty77 committed:
feat: watched bumper auto info

XMinty77 committed:
feat: AutoWatch command

XMinty77 committed:
Merge branch 'EverestAPI:master' into master

XMinty77 committed:
feat: improve AutoWatch command

- support for multiple types at once
- distinct EndAutoWatch command for unwatching
- fixed documentation errors for the command

XMinty77 committed:
feat: improve bumper auto info

- display and draw resultant speed vector
- display y dot product in info hud

XMinty77 committed:
feat: closest spawn point for watched player

draw closest spawn point by distance for auto mode watched player

XMinty77 committed:
Merge branch 'master' into master

DemoJameson commented:
does `UniqueEntityId.GlobalOrPersistent` still need? it has never been read

DemoJameson committed:
fix: info hud width frequent jitter

XMinty7 commented:
> does `UniqueEntityId.GlobalOrPersistent` still need? it has never been read
it's not read but it serves as part of the comparison of UniqueEntityId records, namely for holdables you can transition with and for mod entities that are sometimes persistent and sometimes not
so when it's different it distinguishes persistent entities of the same type vs non persistent entities
I'm not sure because I haven't dealt with the uniqueness tracking system in a while but I think I kept this to solve a bug where the tracker gets confused when you bring a jelly through a transition then go back with it, now you have a duplicate version, and the unique ID alone doesn't distinguish them, but the persistence bool does distinguish them
of course if you duplicate 3+ versions this breaks but I asked about this and it hasn't been reported yet to me an entity or a scenario where you would have 3 or more duplicates in the same room (it's definitely impossible without a glitch in vanilla)

psyGamer commented:
Do you still plan on continuing with this PR?
Also this PRs seems to change various other things not related to the custom info? Did you intend on doing that or did that came from messing something up with git?
Additionally this PR would need to be re-targeted onto `master` due to the various Studio v3 changes. It might make sense to start fresh and copy the relevant things to avoid messy merge conflics and to not have a messy commit history.
